File: Programmers/Printer.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;
// 프로그래머스 level 2 프린터, 맞음!

bool desc(int a, int b){
    return a>b;
}

int solution(vector<int> priorities, int location) {
    int answer = 0;
    vector<int> sorted = priorities;
    queue<pair<int, int>> q;

    sort(sorted.begin(), sorted.end(), desc);

    for(int i=0; i<priorities.size(); i++){
        pair<int, int> p = make_pair(i, priorities[i]);
        q.push(p);
    }


    while (!q.empty()){
        if (!(q.front().second < sorted[0])){  // print
            answer++;
            if (q.front().first == location){
                break;
            }else{
                q.pop();
                sorted.erase(sorted.begin());
            }
        }else{  // not print, go back again
            pair<int, int> front = q.front();
            q.pop();
            q.push(front);
        }
    }

    return answer;
}

int main() {
    int answer;

    answer= solution({1, 1, 9, 1, 1, 1}, 0);

    cout<<answer<<'\n';

    return 0;
}

File: Programmers/Printer.swift
import Foundation

// 프로그래머스 프린터, level 2

struct Queue<T> {
    private var queue: [T] = []
    
    public var count: Int {
        return queue.count
    }
    
    public var front : T {
        return queue[0]
    }
    
    public var isEmpty: Bool {
        return queue.isEmpty
    }
    
    public mutating func enqueue(_ element: T) {
        queue.append(element)
    }
    
    public mutating func dequeue() -> T? {
        return isEmpty ? nil : queue.removeFirst()
    }
}

struct FileInfo {
    var index: Int = -1
    var priority: Int = -1
}


func solution(_ priorities:[Int], _ location:Int) -> Int {
    var q = Queue<FileInfo>()
    var rank = [Int]()
    
    var i = 0
    priorities.forEach {
        q.enqueue(FileInfo(index: i, priority: $0))
        rank.append($0)
        i += 1
    }
    
    rank = rank.sorted(by: >)
    
    var count = 0
    while (!q.isEmpty){
        if q.front.priority < rank[0] {
            if let file = q.dequeue() {
                q.enqueue(file)
            }
            continue
        }
        
        if let file = q.dequeue() {
            count += 1
            if file.index == location {
                break
            }
            
            rank.removeFirst()
        }
    }
    
    return count
}
